tests/test_govuk_frontend.py

```diff
@@ -849,3 +856,156 @@ def test_value_and_error_message(self, question):
+{{ render([{'macro_name': 'test', 'fieldset': None}]) }}""")
+
+        assert template.render() == "<fieldset>bar</fieldset>"
+
+    def test_it_calls_fieldset_with_params(self, env):
+        test_macro = mock.Mock(return_value="foo")
+        test_govuk_fieldset = mock.Mock(
+            side_effect=lambda params, caller: params["fizz"]
+        )
+        template = env.from_string(
+            "{{ render([{'macro_name': 'test', 'fieldset': {'fizz': 'buzz'}}]) }}"
+        )
+        assert template.render(test=test_macro, govukFieldset=test_govuk_fieldset) \
+            == "buzz"
+        assert test_govuk_fieldset.call_args_list == [
+            mock.call({"fizz": "buzz"}, caller=mock.ANY)
+        ]
+
+    def test_macro_call_params_can_contain_macro_calls_in_html_key(self, env):
+        template = env.from_string("""{% macro test() -%}
+<p>foo</p>
+{%- endmacro %}
+{% macro wrapper(params) -%}
+<div>{{ params.html | safe }}</div>
+{%- endmacro %}
+{{ render([{'macro_name': 'wrapper', 'params': {'html': {'macro_name': 'test'}}}]) }}""")
+        assert template.render() == "<div><p>foo</p></div>"
+
+    def test_macro_call_params_can_contain_macro_calls_in_html_key_even_if_nested(self, env):
+        template = env.from_string("""{% macro test() -%}
+<p>foo</p>
+{%- endmacro %}
+{% macro wrapper(params) -%}
+{% for item in params['items'] -%}
+{% if item.conditional -%}
+<div>{{ item.conditional.html | safe }}</div>
+{%- endif %}
+{%- endfor %}
+{%- endmacro %}
+{{ render([{
+    'macro_name': 'wrapper',
+    'params': {'items': [{'conditional': {'html': {'macro_name': 'test'}}}]},
+}]) }}""")
+        assert template.render() == "<div><p>foo</p></div>"
+
+
+class TestRenderQuestion:
+
+    def test_it_calls_from_question_and_render(self):
+        ctx = mock.Mock()
+        question = mock.Mock()
+        data = mock.Mock()
+        errors = mock.Mock()
+
+        with mock.patch("dmcontent.govuk_frontend.from_question") as from_question:
+            with mock.patch("dmcontent.govuk_frontend.render") as render:
+                # with default args
+                render_question(ctx, question)
+
+                assert from_question.call_args == mock.call(question, None, None)
+                assert render.call_args == mock.call(
+                    ctx,
+                    from_question(question, None, None),
+                    question=question
+                )
+
+                # with args and kwargs
+                render_question(ctx, question, data, errors, some_kwarg="foobar")
+
+                assert from_question.call_args == mock.call(
+                    question, data, errors, some_kwarg="foobar"
+                )
+                assert render.call_args == mock.call(
+                    ctx,
+                    from_question(question, data, errors, some_kwarg="foobar"),
+                    question=question
+                )
+
+    def test_it_renders_question_advice(self):
+        pass
```

Comment:
What's this for?

Reply:
😱 I forgot to write that! Thanks! I wanted to test that it recreates [question advice functionality from `show_question.html`](https://github.com/alphagov/digitalmarketplace-brief-responses-frontend/blob/master/app/templates/macros/show_question.html#L27) correctly

Reply:
I figured it was either that, or there was some inheritance magic elsewhere that I was missing 😆

Reply:
Gave the test some body